src/data/heap.lean

```diff
@@ -0,0 +1,227 @@
+lemma tree_sized_zero_is_nil {T : tree α} : size T = 0 ↔ T = nil :=
+begin
+cases T,
+simp only [size, eq_self_iff_true],
+simp,
+end
+
+/-- Minimum distance of node to nil, walking down branches.
+Running time is O(n), where n is the size of the tree. In order to
+optimize the overall implementation, the use of this function should 
+be avoided and instead one should maintain depth / dist_to_nil information
+as part the heap definition - see `insert` documentation for more details.
+-/
+def dist_to_nil : tree α → ℕ
+| nil            := 0
+| (node x t₁ t₂) := min (dist_to_nil t₁) (dist_to_nil t₂) + 1
+
+variables [linear_order α]
+
+/-- Check whether a given tree satisfies the heap invariant. 
+Running time: O(n) - optimal. -/
+def is_heap : tree α → Prop
+| nil            := true
+| (node x t₁ t₂) := ((x : with_top α) ≤ root t₁) ∧
+                    ((x : with_top α) ≤ root t₂) ∧
+                    is_heap t₁ ∧ is_heap t₂
+
+lemma is_heap.left {x : α} {t₁ t₂ : tree α} (h : is_heap (node x t₁ t₂)) : is_heap t₁ :=
+h.2.2.1
+
+lemma is_heap.right {x : α} {t₁ t₂ : tree α} (h : is_heap (node x t₁ t₂)) : is_heap t₂ :=
+h.2.2.2
+
+/-- Lemma: tree with one element is a heap. -/
+lemma is_heap_single (x : α) : is_heap (node x nil nil) :=
+⟨ le_top, le_top, trivial, trivial ⟩
+
+/-- Insertion of a new node with given value to heap.
+Note:
+The running time is O(n log m), where n is the heap size and m is the number 
+of insertion and pop operations in the heap so far (where n is possibly much
+smaller than n). Indeed, one can show that the dist_to_nil is always O(log m):
+For a sequence of m insertions when can show that this follows since we only insert 
+elements in a location with minimal dist_to_nil, and one can also show that pops do 
+not make the situation more problematic.
+
+However, the optimal running time should be O(log n). 
+One way to improve our implementation is as follows:
+1. Improvement from O(n log m) to O(log m): instead of using dist_to_nil computations 
+(which cost O(n) each, where n is the current size of the heap), incorporate a field 
+holding this information as part of the definition of a heap.
+2. Improvement from O(log m) to O(log n): this can be done by making sure that heaps
+retain their balance when a pop operation is carried. In the current implementation,
+when an element is popped, we sift an element along one of the branches of the heap
+(not necessarily the longest) and this can be problematic for retaining balance. 
+A possible solution is to search for a node element with maximum depth in the heap 
+(where depth of node = distance from root), remove it, and add its value 
+to the node we just sifted rather than deleting the said node. To do this efficiently, 
+we need to maintain a depth of a node as part of the definition of a heap. 
+(I think with the correct implementation it is not required to maintain both the depth 
+and the dist_to_nil as fields of a node, perhaps just the depth would be enough.)
```

Comment:
```suggestion
and the `dist_to_nil` as fields of a node, perhaps just the depth would be enough.)
```
Throughout the documentation, could you please use backticks around identifiers?